### PR TITLE
kernel: centralize init of irq stacks

### LINKED ISSUE #7598: move CONFIG_INIT_STACKS logic for IRQ stack to common code
On all arches this is currently done in arch/, usually in some early boot assembly code.
However we don't enable interrupts until we are well into the C domain, and it should be possible to do this somewhere in kernel/init.c, assuming we can standardize the symbol name and kconfig define for the interrupt stack and its size.

### PR DESCRIPTION
**Pull request by tyrantlizard77**
This is now done as soon as we swap into the main thread
entry, since we are no longer on the IRQ stack.

This simplifies arch code, with the caveat that there is
a brief window where interrupts are delivered before
the stack is initialized, and the stack usage of these
interrupts will not be measured.

Fixes: #7598 

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Comment by ioannisg** [comment]
ISR stack is used not only for ISR handling but also for the entire boot sequence for ARM, so it looked essential to initialize it at reset.S. so the "window" is a little larger :)

**Comment by andrewboie** [comment]
> ISR stack is used not only for ISR handling but also for the entire boot sequence for ARM, so it looked essential to initialize it at reset.S. so the "window" is a little larger :)

I'm not sure I understand what you are saying here.

**Comment by andrewboie** [comment]
@ioannisg measuring stack usage of early boot is not what this mechanism is designed to do at all. Some arches don't even use the interrupt stack for early boot. It's really only to measure usage by interrupts.

**Comment by ioannisg** [comment]
> @ioannisg measuring stack usage of early boot is not what this mechanism is designed to do at all. Some arches don't even use the interrupt stack for early boot. It's really only to measure usage by interrupts.

Yep, no objections to the patch, really, I just wanted to stress that the ARM boot sequence does everything using the interrupts stack as the normal (main) stack.

**Comment by andrewboie** [comment]
> Yep, no objections to the patch, really, I just wanted to stress that the ARM boot sequence does everything using the interrupts stack as the normal (main) stack.

I don't understand what point you are trying to make. Can you rephrase or give more detail? What implications does this have? I feel like I'm missing something.

**Comment by ioannisg** [comment]
>I don't understand what point you are trying to make. Can you rephrase or give more detail? What implications does this have? I feel like I'm missing something.

Reading your PR description, I thought you had the impression that the ISR stack is used only for ISR handling. Just to make sure that you won't miss anything, I stressed that the ISR stack in ARM is used as the main stack during kernel boot (and maybe it _did_ make sense to initialize it as early as possible.). I think you have the full picture now.

**Review by dcpleung — APPROVED** [review]

File: kernel/init.c [hunk 1]
```diff
@@ -212,6 +212,19 @@ static void bg_thread_main(void *unused1, void *unused2, void *unused3)
 	static const unsigned int boot_delay;
 #endif
 
+#ifdef CONFIG_INIT_STACKS
+	/* Disable local interrupts on this CPU. Other CPUs haven't been
+	 * started yet.
+	 */
+	int key = arch_irq_lock();
```
Comment (ioannisg) [inline kernel/init.c]: Does it make sense to add the lock-unlock into CONFIG_SMP, then, and save a few bytes of code space for non-SMP builds?

**Review by ioannisg — APPROVED** [review]
Looks OK to me (the ARM and kernel/Init changes)

File: kernel/init.c (hunk 1 above)
Comment (andrewboie) [reply kernel/init.c]: I don't follow you here.
It's essential that local interrupts are not delivered while we are setting the irq stack buffer.
This doesn't have anything to do with SMP.

**Review by wentongwu — APPROVED** [review]

**Review by stephanosio — COMMENTED** [review]
I understand that this is an attempt to make things more SMP-friendly (in particular, the stack initialisation).

One problem I see is, for ARM A and R profile cores, the interrupt stack is not the only per-core stack; there are also ones for each mode (i.e. FIQ, ABT, UND, SVC, SYS).

In such cases, one could argue this central/one-for-all stack initialisation concept breaks down to some degree.

Since stack initialisation is often a highly arch-specific operation, I wonder if it would be better to leave this up to the arch implementation; that would result in a more natural flow and less quirks overall too.

File: kernel/init.c (hunk 1 above)
Comment (stephanosio) [reply kernel/init.c]: Since interrupts are enabled at this point, we must lock-unlock.

**Comment by andrewboie** [comment]
> 
> 
> I understand that this is an attempt to make things more SMP-friendly (in particular, the stack initialisation).
> 
> One problem I see is, for ARM A and R profile cores, the interrupt stack is not the only per-core stack; there are also ones for each mode (i.e. FIQ, ABT, UND, SVC, SYS).
> 
> In such cases, one could argue this central/one-for-all stack initialisation concept breaks down to some degree.
> 
> Since stack initialisation is often a highly arch-specific operation, I wonder if it would be better to leave this up to the arch implementation; that would result in a more natural flow and less quirks overall too.

I want some kind of centralized management, such that in things like the shell module I can dump out the information for all of these stacks.

But you're right, this ignores other platform-specific stacks. x86_64 currently has a set of exception stacks which are also ignored. And looking at the kernel code, hardly anything actually uses the _kernel.cpu[x].irq_stack member.

What I think would be better would probably be something like where all the arch stacks are defined and managed in the arch code, and there's some kind of iteration function to visit all of them. Then we can centrally initialize everything with CONFIG_INIT_STACKS and also dump everything out in the shell.

File: kernel/init.c (hunk 1 above)
Comment (ioannisg) [reply kernel/init.c]: Right, my bad, overlooked that IRQs are enabled at bg_thread_main
